src/v/cluster/controller_snapshot.cc

```diff
@@ -13,11 +13,78 @@
 
 namespace cluster {
 
+namespace controller_snapshot_parts {
+
+template<typename Vec>
+ss::future<> write_vector_async(iobuf& out, Vec t) {
+    if (unlikely(t.size() > std::numeric_limits<serde::serde_size_t>::max())) {
+        throw serde::serde_exception(fmt_with_ctx(
+          ssx::sformat,
+          "serde: vector size {} exceeds serde_size_t",
+          t.size()));
+    }
+    serde::write(out, static_cast<serde::serde_size_t>(t.size()));
+    return ss::do_with(std::move(t), [&out](Vec& t) {
+        return ss::do_for_each(
```

Comment:
does `do_for_each` really guarantee the ordering that we must preserve for a serialization of a vector?

Reply:
I think so. From [the tutorial](https://docs.seastar.io/master/tutorial.html#do_for_each):
```
It accepts a range (or a pair of iterators) and a function body, which it applies to each argument, in order, one by one.
```

Reply:
nice